<a href="https://colab.research.google.com/github/raushan9jnv/Book-Recommender-System/blob/main/Book_Recommender_System.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [62]:
#importing necessary modules
import pandas as pd
import numpy as np

In [63]:
# mounting the data from google drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [64]:
#Loading data
path = '/content/drive/MyDrive/Projects/book-recommender-system/data/'

In [65]:
# load dataset
books = pd.read_csv(path + 'Books.csv')
users=pd.read_csv(path + 'Users.csv')
ratings = pd.read_csv(path + 'Ratings.csv')

<ipython-input-65-99e76f71b468>:2: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  books = pd.read_csv(path + 'Books.csv')


In [66]:
#book dataset
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [67]:
#users dataset
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [68]:
#ratings dataset
ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [69]:
 #shape of data
 print(books.shape)
 print(users.shape)
 print(ratings.shape)

(271360, 8)
(278858, 3)
(1149780, 3)


### Check for null values

In [70]:
books.isnull().sum()

ISBN                   0
Book-Title             0
Book-Author            1
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

In [71]:
users.isnull().sum()

User-ID          0
Location         0
Age         110762
dtype: int64

In [72]:
ratings.isnull().sum()

User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

### Chek for duplicates value

In [73]:
books.duplicated().sum()

0

In [74]:
users.duplicated().sum()

0

In [75]:
ratings.duplicated().sum()

0

## Popularity based Recomender System

We will display top 50 books with Highest average Ratings but we will consider only those books which have got a  minimum of 250 vote.

In [76]:
# Merge ratings and books dataset
rating_with_names = ratings.merge(books, on ="ISBN")

#shape of merge data
rating_with_names.shape

(1031136, 10)

 here the ratings number become 1031136 from 1149780 reason is some rating's books are not present in book dataset so some rows decreases.

In [77]:
# this dataset there are multiple books whose name are same  but there isbn are same means harry potter books are avaiable with three different isbn number.
num_rating_df = rating_with_names.groupby("Book-Title").count()['Book-Rating'].reset_index()
num_rating_df.rename(columns = {'Book-Rating': 'num_ratings'}, inplace=True)
num_rating_df

Book-Title  num_ratings
0        A Light in the Storm: The Civil War Diary of ...            4
1                                   Always Have Popsicles            1
2                    Apple Magic (The Collector's series)            1
3        Ask Lily (Young Women of Faith: Lily Series, ...            1
4        Beyond IBM: Leadership Marketing and Finance ...            1
...                                                   ...          ...
241066                                      Ã?Â?lpiraten.            2
241067                     Ã?Â?rger mit Produkt X. Roman.            4
241068                                Ã?Â?sterlich leben.            1
241069                              Ã?Â?stlich der Berge.            3
241070                                  Ã?Â?thique en toc            2

[241071 rows x 2 columns]

In [78]:
#averge rating
avg_rating_df = rating_with_names.groupby("Book-Title").mean()['Book-Rating'].reset_index()
avg_rating_df.rename(columns = {'Book-Rating': 'avg_ratings'}, inplace=True)
avg_rating_df

<ipython-input-78-eca862e29174>:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  avg_rating_df = rating_with_names.groupby("Book-Title").mean()['Book-Rating'].reset_index()


Book-Title  avg_ratings
0        A Light in the Storm: The Civil War Diary of ...     2.250000
1                                   Always Have Popsicles     0.000000
2                    Apple Magic (The Collector's series)     0.000000
3        Ask Lily (Young Women of Faith: Lily Series, ...     8.000000
4        Beyond IBM: Leadership Marketing and Finance ...     0.000000
...                                                   ...          ...
241066                                      Ã?Â?lpiraten.     0.000000
241067                     Ã?Â?rger mit Produkt X. Roman.     5.250000
241068                                Ã?Â?sterlich leben.     7.000000
241069                              Ã?Â?stlich der Berge.     2.666667
241070                                  Ã?Â?thique en toc     4.000000

[241071 rows x 2 columns]

In [79]:
#merge num_rating_df and avg_rating_df
popular_df = num_rating_df.merge(avg_rating_df, on = 'Book-Title')
popular_df

Book-Title  num_ratings  \
0        A Light in the Storm: The Civil War Diary of ...            4   
1                                   Always Have Popsicles            1   
2                    Apple Magic (The Collector's series)            1   
3        Ask Lily (Young Women of Faith: Lily Series, ...            1   
4        Beyond IBM: Leadership Marketing and Finance ...            1   
...                                                   ...          ...   
241066                                      Ã?Â?lpiraten.            2   
241067                     Ã?Â?rger mit Produkt X. Roman.            4   
241068                                Ã?Â?sterlich leben.            1   
241069                              Ã?Â?stlich der Berge.            3   
241070                                  Ã?Â?thique en toc            2   

        avg_ratings  
0          2.250000  
1          0.000000  
2          0.000000  
3          8.000000  
4          0.000000  
...             ...  
241066     0.000000  
241067     5.250000  
241068     7.000000  
241069     2.666667  
241070     4.000000  

[241071 rows x 3 columns]

In [80]:
# ratings greater or equal to 250 and select top 50
popular_df = popular_df[popular_df['num_ratings'] >= 250].sort_values('avg_ratings', ascending = False).head(50)
popular_df

Book-Title  num_ratings  \
80434   Harry Potter and the Prisoner of Azkaban (Book 3)          428   
80422        Harry Potter and the Goblet of Fire (Book 4)          387   
80441      Harry Potter and the Sorcerer's Stone (Book 1)          278   
80426   Harry Potter and the Order of the Phoenix (Boo...          347   
80414    Harry Potter and the Chamber of Secrets (Book 2)          556   
191612  The Hobbit : The Enchanting Prelude to The Lor...          281   
187377  The Fellowship of the Ring (The Lord of the Ri...          368   
80445   Harry Potter and the Sorcerer's Stone (Harry P...          575   
211384     The Two Towers (The Lord of the Rings, Part 2)          260   
219741                              To Kill a Mockingbird          510   
183573                                  The Da Vinci Code          898   
187880                 The Five People You Meet in Heaven          430   
180556                             The Catcher in the Rye          449   
196326                          The Lovely Bones: A Novel         1295   
764                                                  1984          284   
144165                           Prodigal Summer: A Novel          253   
128670                                         Neverwhere          265   
206502                            The Secret Life of Bees          774   
168719  Stupid White Men ...and Other Sorry Excuses fo...          283   
223135  Tuesdays with Morrie: An Old Man, a Young Man,...          493   
204387                The Red Tent (Bestselling Backlist)          723   
191589               The Hitchhiker's Guide to the Galaxy          268   
129379   Nickel and Dimed: On (Not) Getting By in America          335   
93381                                       Into the Wild          252   
63867                                      Fahrenheit 451          409   
74750                           Girl with a Pearl Earring          526   
136145                                          Outlander          283   
233370  Where the Heart Is (Oprah's Book Club (Paperba...          585   
156102                     Seabiscuit: An American Legend          275   
107962                                         Life of Pi          664   
176845                                     The Bean Trees          389   
2281    A Child Called \It\": One Child's Courage to S...          265   
8434                                       ANGELA'S ASHES          279   
76343                                         Good in Bed          490   
64931   Fast Food Nation: The Dark Side of the All-Ame...          321   
12700                                       American Gods          302   
161645                                 Skipping Christmas          322   
105777  Left Behind: A Novel of the Earth's Last Days ...          318   
189551    The Golden Compass (His Dark Materials, Book 1)          336   
181679                                   The Color Purple          314   
160336                               Silence of the Lambs          256   
8752                                          About a Boy          262   
158138                  Seven Up (A Stephanie Plum Novel)          278   
175097  The Alchemist: A Fable About Following Your Dream          266   
80069   Hard Eight : A Stephanie Plum Novel (A Stephan...          269   
170101                       Suzanne's Diary for Nicholas          457   
111073                                  Lord of the Flies          259   
5664                              A Prayer for Owen Meany          413   
212070   The Vampire Lestat (Vampire Chronicles, Book II)          301   
233851       White Oleander : A Novel (Oprah's Book Club)          356   

        avg_ratings  
80434      5.852804  
80422      5.824289  
80441      5.737410  
80426      5.501441  
80414      5.183453  
191612     5.007117  
187377     4.948370  
80445      4.895652  
211384     4.880769  
219741     4.700000  
183573     4.642539  
187880  

In [81]:
popular_df = popular_df.merge(books, on= 'Book-Title').drop_duplicates('Book-Title')[['Book-Title', 'Book-Author','Image-URL-M', 'num_ratings', 'avg_ratings']]

In [82]:
# these are our top 50 books and there details based on Popularity based recommender system using average rating formula
popular_df

Book-Title  \
0    Harry Potter and the Prisoner of Azkaban (Book 3)   
3         Harry Potter and the Goblet of Fire (Book 4)   
5       Harry Potter and the Sorcerer's Stone (Book 1)   
9    Harry Potter and the Order of the Phoenix (Boo...   
13    Harry Potter and the Chamber of Secrets (Book 2)   
16   The Hobbit : The Enchanting Prelude to The Lor...   
17   The Fellowship of the Ring (The Lord of the Ri...   
26   Harry Potter and the Sorcerer's Stone (Harry P...   
28      The Two Towers (The Lord of the Rings, Part 2)   
39                               To Kill a Mockingbird   
47                                   The Da Vinci Code   
53                  The Five People You Meet in Heaven   
55                              The Catcher in the Rye   
62                           The Lovely Bones: A Novel   
63                                                1984   
72                            Prodigal Summer: A Novel   
73                                          Neverwhere   
78                             The Secret Life of Bees   
84   Stupid White Men ...and Other Sorry Excuses fo...   
85   Tuesdays with Morrie: An Old Man, a Young Man,...   
88                 The Red Tent (Bestselling Backlist)   
89                The Hitchhiker's Guide to the Galaxy   
98    Nickel and Dimed: On (Not) Getting By in America   
100                                      Into the Wild   
103                                     Fahrenheit 451   
116                          Girl with a Pearl Earring   
117                                          Outlander   
123  Where the Heart Is (Oprah's Book Club (Paperba...   
124                     Seabiscuit: An American Legend   
126                                         Life of Pi   
130                                     The Bean Trees   
133  A Child Called \It\": One Child's Courage to S...   
135                                     ANGELA'S ASHES   
137                                        Good in Bed   
139  Fast Food Nation: The Dark Side of the All-Ame...   
140                                      American Gods   
141                                 Skipping Christmas   
147  Left Behind: A Novel of the Earth's Last Days ...   
149    The Golden Compass (His Dark Materials, Book 1)   
157                                   The Color Purple   
163                               Silence of the Lambs   
166                                        About a Boy   
173                  Seven Up (A Stephanie Plum Novel)   
175  The Alchemist: A Fable About Following Your Dream   
177  Hard Eight : A Stephanie Plum Novel (A Stephan...   
182                       Suzanne's Diary for Nicholas   
186                                  Lord of the Flies   
192                            A Prayer for Owen Meany   
194   The Vampire Lestat (Vampire Chronicles, Book II)   
195       White Oleander : A Novel (Oprah's Book Club)   

                Book-Author  \
0             J. K. Rowling   
3             J. K. Rowling   
5             J. K. Rowling   
9             J. K. Rowling   
13            J. K. Rowling   
16           J.R.R. TOLKIEN   
17           J.R.R. TOLKIEN   
26            J. K. Rowling   
28           J.R.R. TOLKIEN   
39               Harper Lee   
47                Dan Brown   
53              Mitch Albom   
55            J.D. Salinger   
62             Alice Sebold   
63            George Orwell   
72       Barbara Kingsolver   
73              Neil Gaiman   
78            Sue Monk Kidd   
84            Michael Moore   
85              MITCH ALBOM   
88            Anita Diamant   
89            Douglas Adams   
98       Barbara Ehrenreich   
100            Jon Krakauer   
103            Ray Bradbury   
116         Tracy Chevalier   
117          DIANA GABALDON   
123            Billie Letts   
124       LAURA HILLENBRAND   
126             Yann Martel   
130      Barbara Kingsolver   
133             Dave Pelzer   
135           Frank McCourt   
137         Jennifer Weiner   


# Collaborative filtering based recommender system

after working on this data, I relise, in this we are not considering all users  and books in model building as it was not giving accurate predicitions.
- So after error and trial method (EXPERIMENT) i come to conclusion that we are considering only those user who have given minimum 200 reviews as they are experienced and we considering only those books which have minimum 50 reviws given by people. so by this we negelect unpopular book which are known by very less people. to make our model/system more intelligent.

USER > 200 REVIEW GIVEN

BOOKS, REVIEW BY >= 50 PEOPLE

In [83]:
#filtering based on user
x = rating_with_names.groupby('User-ID').count()['Book-Rating'] > 200
user_200 = x[x].index
user_200

Int64Index([   254,   2276,   2766,   2977,   3363,   4017,   4385,   6251,
              6323,   6543,
            ...
            271705, 273979, 274004, 274061, 274301, 274308, 275970, 277427,
            277639, 278418],
           dtype='int64', name='User-ID', length=811)

In [84]:
filtered_rating = rating_with_names[rating_with_names['User-ID'].isin(user_200)]

In [85]:
#filtering based on books

y =filtered_rating.groupby('Book-Title').count()['Book-Rating']>=50
famous_books = y[y].index
famous_books

Index(['1984', '1st to Die: A Novel', '2nd Chance', '4 Blondes',
       'A Bend in the Road', 'A Case of Need',
       'A Child Called \It\": One Child's Courage to Survive"',
       'A Civil Action', 'A Day Late and a Dollar Short', 'A Fine Balance',
       ...
       'Winter Solstice', 'Wish You Well', 'Without Remorse',
       'Wizard and Glass (The Dark Tower, Book 4)', 'Wuthering Heights',
       'Year of Wonders', 'You Belong To Me',
       'Zen and the Art of Motorcycle Maintenance: An Inquiry into Values',
       'Zoya', '\O\" Is for Outlaw"'],
      dtype='object', name='Book-Title', length=706)

In [86]:
final_ratings = filtered_rating[filtered_rating['Book-Title'].isin(famous_books)]
final_ratings

User-ID        ISBN  Book-Rating              Book-Title  \
63        278418  0446520802            0            The Notebook   
65          3363  0446520802            0            The Notebook   
66          7158  0446520802           10            The Notebook   
69         11676  0446520802           10            The Notebook   
74         23768  0446520802            6            The Notebook   
...          ...         ...          ...                     ...   
1026724   266865  0531001725           10  The Catcher in the Rye   
1027923   269566  0670809381            0                  Echoes   
1028777   271284  0440910927            0           The Rainmaker   
1029070   271705  B0001PIOX4            0          Fahrenheit 451   
1030868   275970  1586210661            9  Me Talk Pretty One Day   

                   Book-Author Year-Of-Publication                Publisher  \
63             Nicholas Sparks                1996             Warner Books   
65             Nicholas Sparks                1996             Warner Books   
66             Nicholas Sparks                1996             Warner Books   
69             Nicholas Sparks                1996             Warner Books   
74             Nicholas Sparks                1996             Warner Books   
...                        ...                 ...                      ...   
1026724  Jerome David Salinger                1973   Scholastic Library Pub   
1027923           Maeve Binchy                1986              Penguin USA   
1028777           John Grisham                1995                   Island   
1029070           Ray Bradbury                1993     Simon &amp; Schuster   
1030868          David Sedaris                2001  Time Warner Audio Major   

                                               Image-URL-S  \
63       http://images.amazon.com/images/P/0446520802.0...   
65       http://images.amazon.com/images/P/0446520802.0...   
66       http://images.amazon.com/images/P/0446520802.0...   
69       http://images.amazon.com/images/P/0446520802.0...   
74       http://images.amazon.com/images/P/0446520802.0...   
...                                                    ...   
1026724  http://images.amazon.com/images/P/0531001725.0...   
1027923  http://images.amazon.com/images/P/0670809381.0...   
1028777  http://images.amazon.com/images/P/0440910927.0...   
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...   
1030868  http://images.amazon.com/images/P/1586210661.0...   

                                               Image-URL-M  \
63       http://images.amazon.com/images/P/0446520802.0...   
65       http://images.amazon.com/images/P/0446520802.0...   
66       http://images.amazon.com/images/P/0446520802.0...   
69       http://images.amazon.com/images/P/0446520802.0...   
74       http://images.amazon.com/images/P/0446520802.0...   
...                                                    ...   
1026724  http://images.amazon.com/images/P/0531001725.0...   
1027923  http://images.amazon.com/images/P/0670809381.0...   
1028777  http://images.amazon.com/images/P/0440910927.0...   
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...   
1030868  http://images.amazon.com/images/P/1586210661.0...   

                                               Image-URL-L  
63       http://images.amazon.com/images/P/0446520802.0...  
65       http://images.amazon.com/images/P/0446520802.0...  
66       http://images.amazon.com/images/P/0446520802.0...  
69       http://images.amazon.com/images/P/0446520802.0...  
74       http://images.amazon.com/images/P/0446520802.0...  
...                                                    ...  
1026724  http://images.amazon.com/images/P/0531001725.0...  
1027923  http://images.amazon.com/images/P/0670809381.0...  
1028777  http://images.amazon.com/images/P/0440910927.0...  
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...  
1030868  http://images.amazon.com/images/P/1586210661.0...  

[58586 r

In [87]:
final_ratings.drop_duplicates()

User-ID        ISBN  Book-Rating              Book-Title  \
63        278418  0446520802            0            The Notebook   
65          3363  0446520802            0            The Notebook   
66          7158  0446520802           10            The Notebook   
69         11676  0446520802           10            The Notebook   
74         23768  0446520802            6            The Notebook   
...          ...         ...          ...                     ...   
1026724   266865  0531001725           10  The Catcher in the Rye   
1027923   269566  0670809381            0                  Echoes   
1028777   271284  0440910927            0           The Rainmaker   
1029070   271705  B0001PIOX4            0          Fahrenheit 451   
1030868   275970  1586210661            9  Me Talk Pretty One Day   

                   Book-Author Year-Of-Publication                Publisher  \
63             Nicholas Sparks                1996             Warner Books   
65             Nicholas Sparks                1996             Warner Books   
66             Nicholas Sparks                1996             Warner Books   
69             Nicholas Sparks                1996             Warner Books   
74             Nicholas Sparks                1996             Warner Books   
...                        ...                 ...                      ...   
1026724  Jerome David Salinger                1973   Scholastic Library Pub   
1027923           Maeve Binchy                1986              Penguin USA   
1028777           John Grisham                1995                   Island   
1029070           Ray Bradbury                1993     Simon &amp; Schuster   
1030868          David Sedaris                2001  Time Warner Audio Major   

                                               Image-URL-S  \
63       http://images.amazon.com/images/P/0446520802.0...   
65       http://images.amazon.com/images/P/0446520802.0...   
66       http://images.amazon.com/images/P/0446520802.0...   
69       http://images.amazon.com/images/P/0446520802.0...   
74       http://images.amazon.com/images/P/0446520802.0...   
...                                                    ...   
1026724  http://images.amazon.com/images/P/0531001725.0...   
1027923  http://images.amazon.com/images/P/0670809381.0...   
1028777  http://images.amazon.com/images/P/0440910927.0...   
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...   
1030868  http://images.amazon.com/images/P/1586210661.0...   

                                               Image-URL-M  \
63       http://images.amazon.com/images/P/0446520802.0...   
65       http://images.amazon.com/images/P/0446520802.0...   
66       http://images.amazon.com/images/P/0446520802.0...   
69       http://images.amazon.com/images/P/0446520802.0...   
74       http://images.amazon.com/images/P/0446520802.0...   
...                                                    ...   
1026724  http://images.amazon.com/images/P/0531001725.0...   
1027923  http://images.amazon.com/images/P/0670809381.0...   
1028777  http://images.amazon.com/images/P/0440910927.0...   
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...   
1030868  http://images.amazon.com/images/P/1586210661.0...   

                                               Image-URL-L  
63       http://images.amazon.com/images/P/0446520802.0...  
65       http://images.amazon.com/images/P/0446520802.0...  
66       http://images.amazon.com/images/P/0446520802.0...  
69       http://images.amazon.com/images/P/0446520802.0...  
74       http://images.amazon.com/images/P/0446520802.0...  
...                                                    ...  
1026724  http://images.amazon.com/images/P/0531001725.0...  
1027923  http://images.amazon.com/images/P/0670809381.0...  
1028777  http://images.amazon.com/images/P/0440910927.0...  
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...  
1030868  http://images.amazon.com/images/P/1586210661.0...  

[58586 r

### pivot table

In [88]:
#these are the people who read more than 200 books and books are minimum reviewed by 50 peoples. represented by pivot table
pt = final_ratings.pivot_table(index = 'Book-Title', columns = 'User-ID', values = 'Book-Rating')
pt

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN    10.0     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     0.0     NaN     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     NaN     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     0.0     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   NaN  ...    10.0   
1st to Die: A Novel                                    9.0  ...     NaN   
2nd Chance                                             0.0  ...     NaN   
4 Blondes                                              NaN  ...     NaN   
A Bend in the Road                                     NaN  ...     NaN   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     NaN   
You Belong To Me                                       NaN  ...     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN  ...     NaN   
Zoya                                                   NaN  ...     NaN   
\O\" Is for Outlaw"                                    

there is 706 famous books and 810 trusted users

In [89]:
pt.fillna(0, inplace = True)
pt

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   0.0  ...    10.0   
1st to Die: A Novel                                    9.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
A Bend in the Road                                     0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

Now calculate  eucliden distances

In [90]:
from  sklearn.metrics.pairwise import cosine_similarity

In [91]:
#calculate cosine similarity of all rows with oother rows, it will give 2D numpy array
similarity_scores = cosine_similarity(pt)
similarity_scores

array([[1.        , 0.10255025, 0.01220856, ..., 0.12110367, 0.07347567,
        0.04316046],
       [0.10255025, 1.        , 0.2364573 , ..., 0.07446129, 0.16773875,
        0.14263397],
       [0.01220856, 0.2364573 , 1.        , ..., 0.04558758, 0.04938579,
        0.10796119],
       ...,
       [0.12110367, 0.07446129, 0.04558758, ..., 1.        , 0.07085128,
        0.0196177 ],
       [0.07347567, 0.16773875, 0.04938579, ..., 0.07085128, 1.        ,
        0.10602962],
       [0.04316046, 0.14263397, 0.10796119, ..., 0.0196177 , 0.10602962,
        1.        ]])

In [92]:
similarity_scores.shape

(706, 706)

In [93]:
similarity_scores[0]

array([1.        , 0.10255025, 0.01220856, 0.        , 0.05367224,
       0.02774901, 0.08216491, 0.13732869, 0.03261686, 0.03667591,
       0.02322418, 0.06766487, 0.02083978, 0.09673735, 0.13388865,
       0.08303112, 0.11153543, 0.05100411, 0.02517784, 0.11706383,
       0.        , 0.14333793, 0.07847534, 0.06150451, 0.08723968,
       0.        , 0.07009814, 0.13658681, 0.07600328, 0.12167134,
       0.00768046, 0.01473221, 0.        , 0.07965814, 0.04522617,
       0.01556271, 0.09495938, 0.0182307 , 0.02610465, 0.07984012,
       0.11679969, 0.0569124 , 0.08354155, 0.08471898, 0.08785938,
       0.05491435, 0.0548505 , 0.27026514, 0.09779123, 0.06016046,
       0.08958835, 0.06748675, 0.        , 0.04468098, 0.01920872,
       0.        , 0.05629067, 0.00557964, 0.07877059, 0.05219479,
       0.18908177, 0.        , 0.01240656, 0.02984572, 0.04279502,
       0.12680125, 0.16566735, 0.        , 0.13357242, 0.06615478,
       0.        , 0.        , 0.        , 0.10968075, 0.02806

In [94]:
def recommend(book_name):
  #index fetch
  index = np.where(pt.index == book_name)[0][0]
  similar_items = sorted(list(enumerate(similarity_scores[index])),key = lambda x:x[1], reverse = True)[1:6]

 # for print only books name

  # for i in similar_items:
  #   print(pt.index[i[0]])

# print books author image url

  data =[]
  for i in similar_items:
    item = []
    temp_df = books[books['Book-Title'] == pt.index[i[0]]]
    item.extend(list(temp_df.drop_duplicates('Book-Title')['Book-Title'].values)) # don't use APPEND with LIST use EXTEND because list become 2D array after append.
    item.extend(list(temp_df.drop_duplicates('Book-Title')['Book-Author'].values))
    item.extend(list(temp_df.drop_duplicates('Book-Title')['Image-URL-M'].values))

    data.append(item)

  return data

In [95]:
recommend('1984')

[['Animal Farm',
  'George Orwell',
  'http://images.amazon.com/images/P/0451526341.01.MZZZZZZZ.jpg'],
 ["The Handmaid's Tale",
  'Margaret Atwood',
  'http://images.amazon.com/images/P/0449212602.01.MZZZZZZZ.jpg'],
 ['Brave New World',
  'Aldous Huxley',
  'http://images.amazon.com/images/P/0060809833.01.MZZZZZZZ.jpg'],
 ['The Vampire Lestat (Vampire Chronicles, Book II)',
  'ANNE RICE',
  'http://images.amazon.com/images/P/0345313860.01.MZZZZZZZ.jpg'],
 ['The Hours : A Novel',
  'Michael Cunningham',
  'http://images.amazon.com/images/P/0312243022.01.MZZZZZZZ.jpg']]

In [96]:
recommend('Message in a Bottle')  # a romatic book, all recommend all also same

[['Nights in Rodanthe',
  'Nicholas Sparks',
  'http://images.amazon.com/images/P/0446531332.01.MZZZZZZZ.jpg'],
 ['The Mulberry Tree',
  'Jude Deveraux',
  'http://images.amazon.com/images/P/0743437640.01.MZZZZZZZ.jpg'],
 ['A Walk to Remember',
  'Nicholas Sparks',
  'http://images.amazon.com/images/P/0446608955.01.MZZZZZZZ.jpg'],
 ["River's End",
  'Nora Roberts',
  'http://images.amazon.com/images/P/0515127833.01.MZZZZZZZ.jpg'],
 ['Nightmares &amp; Dreamscapes',
  'Stephen King',
  'http://images.amazon.com/images/P/0451180232.01.MZZZZZZZ.jpg']]

In [97]:
recommend('The Notebook')

[['A Walk to Remember',
  'Nicholas Sparks',
  'http://images.amazon.com/images/P/0446608955.01.MZZZZZZZ.jpg'],
 ['The Rescue',
  'Nicholas Sparks',
  'http://images.amazon.com/images/P/0446610399.01.MZZZZZZZ.jpg'],
 ['One Door Away from Heaven',
  'Dean R. Koontz',
  'http://images.amazon.com/images/P/0553582755.01.MZZZZZZZ.jpg'],
 ['Toxin',
  'Robin Cook',
  'http://images.amazon.com/images/P/0425166619.01.MZZZZZZZ.jpg'],
 ['The Five People You Meet in Heaven',
  'Mitch Albom',
  'http://images.amazon.com/images/P/0786868716.01.MZZZZZZZ.jpg']]

In [98]:
recommend('Animal Farm')  # a famous book on politics

[['1984',
  'George Orwell',
  'http://images.amazon.com/images/P/0451524934.01.MZZZZZZZ.jpg'],
 ['Angus, Thongs and Full-Frontal Snogging: Confessions of Georgia Nicolson',
  'Louise Rennison',
  'http://images.amazon.com/images/P/0064472272.01.MZZZZZZZ.jpg'],
 ['Midnight',
  'Dean R. Koontz',
  'http://images.amazon.com/images/P/0425118703.01.MZZZZZZZ.jpg'],
 ['Second Nature',
  'Alice Hoffman',
  'http://images.amazon.com/images/P/0399139087.01.MZZZZZZZ.jpg'],
 ['Call of the Wild',
  'Jack London',
  'http://images.amazon.com/images/P/1559029838.01.MZZZZZZZ.jpg']]

In [99]:
recommend('The Da Vinci Code')

[['Angels &amp; Demons',
  'Dan Brown',
  'http://images.amazon.com/images/P/0671027360.01.MZZZZZZZ.jpg'],
 ['Touching Evil',
  'Kay Hooper',
  'http://images.amazon.com/images/P/0553583441.01.MZZZZZZZ.jpg'],
 ['Saving Faith',
  'David Baldacci',
  'http://images.amazon.com/images/P/0446608890.01.MZZZZZZZ.jpg'],
 ["The Sweet Potato Queens' Book of Love",
  'JILL CONNER BROWNE',
  'http://images.amazon.com/images/P/0609804138.01.MZZZZZZZ.jpg'],
 ['Middlesex: A Novel',
  'Jeffrey Eugenides',
  'http://images.amazon.com/images/P/0312422156.01.MZZZZZZZ.jpg']]

In [100]:
recommend('Harry Potter and the Goblet of Fire (Book 4)')

[['Harry Potter and the Prisoner of Azkaban (Book 3)',
  'J. K. Rowling',
  'http://images.amazon.com/images/P/0439136350.01.MZZZZZZZ.jpg'],
 ['Harry Potter and the Chamber of Secrets (Book 2)',
  'J. K. Rowling',
  'http://images.amazon.com/images/P/0439064872.01.MZZZZZZZ.jpg'],
 ['Harry Potter and the Order of the Phoenix (Book 5)',
  'J. K. Rowling',
  'http://images.amazon.com/images/P/043935806X.01.MZZZZZZZ.jpg'],
 ["Harry Potter and the Sorcerer's Stone (Book 1)",
  'J. K. Rowling',
  'http://images.amazon.com/images/P/0590353403.01.MZZZZZZZ.jpg'],
 ["Harry Potter and the Sorcerer's Stone (Harry Potter (Paperback))",
  'J. K. Rowling',
  'http://images.amazon.com/images/P/059035342X.01.MZZZZZZZ.jpg']]

Book name for check

- Message in a Bottle - A romentic book
- The Notebook - A romentic book
- 1984
- the da vinci code
- The hobbit: The enchanting prelude to the lord of the rings
- The Fellowship of the RIng(The Lord of the Rings, Part 1)
- Harry Potter and the Chamber of the Secrets(Book 2)
- Harry Potter and the Goblet of Fire (Book 4)


Now deploy it on website
- we use streamlit framework

In [108]:
import pickle
pickle.dump(popular_df, open('/content/drive/MyDrive/Projects/book-recommender-system/model/popular.pkl','wb'))

In [102]:
print(pd. __version__)

1.5.3


### Dump Collabrative filtering Model for Deployment on Streamlit

It uses three variables so we have to export that i.e. pt, books, similarity_score

In [109]:
pickle.dump(pt, open('/content/drive/MyDrive/Projects/book-recommender-system/model/pt.pkl','wb'))
pickle.dump(books, open('/content/drive/MyDrive/Projects/book-recommender-system/model/books.pkl','wb'))
pickle.dump(similarity_scores, open('/content/drive/MyDrive/Projects/book-recommender-system/model/similarity_scores.pkl','wb'))